sdks/python/apache_beam/runners/sdf_common.py

```diff
@@ -0,0 +1,161 @@
+      return SplittableParDo(ptransform)
+    else:
+      return ptransform
+
+
+class SplittableParDo(PTransform):
+  """A transform that processes a PCollection using a Splittable DoFn."""
+
+  def __init__(self, ptransform):
+    assert isinstance(ptransform, ParDo)
+    self._ptransform = ptransform
+
+  def expand(self, pcoll):
+    sdf = self._ptransform.fn
+    signature = DoFnSignature(sdf)
+    invoker = DoFnInvoker.create_invoker(signature, process_invocation=False)
+
+    element_coder = typecoders.registry.get_coder(pcoll.element_type)
+    restriction_coder = invoker.invoke_restriction_coder()
+
+    keyed_elements = (pcoll
+                      | 'pair' >> ParDo(PairWithRestrictionFn(sdf))
+                      | 'split' >> ParDo(SplitRestrictionFn(sdf))
+                      | 'explode' >> ParDo(ExplodeWindowsFn())
+                      | 'random' >> ParDo(RandomUniqueKeyFn()))
+
+    return keyed_elements | ProcessKeyedElements(
+        sdf, element_coder, restriction_coder,
+        pcoll.windowing, self._ptransform.args, self._ptransform.kwargs)
+
+
+class ElementAndRestriction(object):
+  """A holder for an element and a restriction."""
+
+  def __init__(self, element, restriction):
+    self.element = element
+    self.restriction = restriction
+
+
+class PairWithRestrictionFn(beam.DoFn):
+  """A transform that pairs each element with a restriction."""
+
+  def __init__(self, do_fn):
+    signature = DoFnSignature(do_fn)
+    self._invoker = DoFnInvoker.create_invoker(
+        signature, process_invocation=False)
+
+  def process(self, element, window=beam.DoFn.WindowParam, *args, **kwargs):
+    initial_restriction = self._invoker.invoke_initial_restriction(element)
+    yield ElementAndRestriction(element, initial_restriction)
+
+
+class SplitRestrictionFn(beam.DoFn):
+  """A transform that perform initial splitting of Splittable DoFn inputs."""
+
+  def __init__(self, do_fn):
+    signature = DoFnSignature(do_fn)
+    self._invoker = DoFnInvoker.create_invoker(
+        signature, process_invocation=False)
+
+  def process(self, element_and_restriction, *args, **kwargs):
+    element = element_and_restriction.element
+    restriction = element_and_restriction.restriction
+    restriction_parts = self._invoker.invoke_split(
+        element,
+        restriction)
+    for part in restriction_parts:
+      yield ElementAndRestriction(element, part)
+
+
+class ExplodeWindowsFn(beam.DoFn):
+  """A transform that forces the runner to explode windows.
+
+  This is done to make sure that Splittable DoFn proceses an element for each of
+  the windows that element belongs to.
+  """
+
+  def process(self, element, window=beam.DoFn.WindowParam, *args, **kwargs):
+    yield element
+
+
+class RandomUniqueKeyFn(beam.DoFn):
+  """A transform that assigns a unique key to each element."""
+
+  def process(self, element, window=beam.DoFn.WindowParam, *args, **kwargs):
+    yield (uuid.uuid4().bytes, element)
```

Comment:
<!--thread_id:cc_151534745_t; commit:7549b5c2ebe2ae47af9066eaf97364a27e828ab5; resolved:0-->
<!--section:context-quote-->
> **chamikaramj** wrote:
> Added TODOs to here and iobase._WriteBundleDoFn. I think collisions are extremely rare for uuid.uuid4() though. Also, added an assertion to force a failure here if a collision is detected.

<!--section:body-->
I don't see this assertion?  I agree that they should be rare, but a comment would help anyone reading this in the future understand this assumption.